### PR TITLE
StereoPannerNode DOM

### PR DESCRIPTION
**Pull request by PurpleHairEngineer**
This is the DOM implementation for StereoPannerNode. Backend implemented in https://github.com/servo/media/pull/243.

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23281)

### CODE REVIEW

File: components/script/dom/stereopannernode.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    pub fn new_inherited(
+        window: &Window,
+        context: &BaseAudioContext,
+        options: &StereoPannerOptions,
+    ) -> Fallible<StereoPannerNode> {
+        let node_options =
+            options
+                .parent
+                .unwrap_or(2, ChannelCountMode::Max, ChannelInterpretation::Speakers);
+        let source_node = AudioScheduledSourceNode::new_inherited(
```
Comment (Manishearth) [inline components/script/dom/stereopannernode.rs]: You need to validate based on constraints here, see https://webaudio.github.io/web-audio-api/#audionode-channelcount-constraints , https://webaudio.github.io/web-audio-api/#dom-audionode-channelcountmode : if the channel count mode is Max or the channel count is greater than two, return a NotSupported Error

You'll also need to add similar constraints on SetChannelCount and SetChannelCountMode on audionode.rs

File: components/script/dom/stereopannernode.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    #[allow(unrooted_must_root)]
+    pub fn new_inherited(
+        window: &Window,
+        context: &BaseAudioContext,
+        options: &StereoPannerOptions,
+    ) -> Fallible<StereoPannerNode> {
+        let node_options =
+            options
+                .parent
+                .unwrap_or(2, ChannelCountMode::Max, ChannelInterpretation::Speakers);
```
Comment (Manishearth) [inline components/script/dom/stereopannernode.rs]: the default mode is ClampedMax here

**Comment by PurpleHairEngineer** [comment]
@Manishearth Okay, I believe I added the constraints appropriately.

**Comment by PurpleHairEngineer** [comment]
@Manishearth I see that there are build errors for this, but a lot of them seem to be related to the fact that it can't find stereo_panner_node in servo_media. That's being tracked in a different pull request, https://github.com/servo/media/pull/243 . Is there anything else I can do for this one before I get that one working?

**Comment by Manishearth** [comment]
Nah, get that one working first. There's a way to locally redirect dependencies but I'd rather just merge that first and look at this one later.

**Comment by PurpleHairEngineer** [comment]
@Manishearth Now that https://github.com/servo/media/pull/243 has been merged, what is the best way to update this branch to reflect those changes? Should I pull the master branch, or something else?

**Comment by Manishearth** [comment]
`./mach cargo update -p servo-media` should be enough. Commit the changes it causes.

Also, you should first rebase and squash this (there will be a small merge conflict that needs fixing) so you won't have the same problem you had in the other PR.

Ideally if you can, make the commit that updates servo-media a separate commit _before_ the main one (it's easier to rebase this in case of lockfile issues), but don't worry about that if you don't know how.

**Comment by PurpleHairEngineer** [comment]
@Manishearth You are saying I should squash before running the `/mach cargo update -p servo-media` command? When I tried to run `git rebase -i HEAD~5`, it showed my commits at the top and then almost 400 commits after that. Are you saying I should squash them all?

**Comment by Manishearth** [comment]
> `git rebase -i HEAD~5`

Run `git fetch origin` and then `git rebase -i origin/master`. Only squash your commits, no other commits should show up. I can do it for you if you need.

Yes, squash first.

**Comment by PurpleHairEngineer** [comment]
@Manishearth I ran `git rebase -i origin/master`, but in the editor, it is showing my 4 commits on top followed by about 390 others. Can you do it for me? Sorry about this.

**Comment by jdm** [comment]
If you remove all of the entries and close the editor, the rebase will be cancelled without making any changes.

**Comment by PurpleHairEngineer** [comment]
@jdm Thanks for the tip! Sorry, I don't have a lot of experience using git before this project.

**Comment by Manishearth** [comment]
On a bus right now, will  push later. Is `origin` the servo repo or your fork? If it's your fork, do `git remote add upstream https://github.com/servo/servo.git` and then use `git rebase -i upstream/master`.

This worked for me locally.

**Comment by PurpleHairEngineer** [comment]
@Manishearth `origin` is the fork (https://github.com/josephhutch/servo). Thanks! I had to run `git fetch` in between those two commands, but I think it worked!

**Comment by Manishearth** [comment]
Pushed. I also included the update of servo-media.

Run the following to get the update locally

```
git fetch origin
git checkout origin/StereoPannerNode
git checkout -B StereoPannerNode
git push origin StereoPannerNode -u
```

**Comment by PurpleHairEngineer** [comment]
@Manishearth I'm getting the following error:

```
error[E0599]: no variant or associated item named `StereoPannerNode` found for type `dom::bindings::codegen::InheritTypes::AudioNodeTypeId` in the current scope
   --> components\script\dom\audionode.rs:288:59
    |
288 |             EventTargetTypeId::AudioNode(AudioNodeTypeId::StereoPannerNode) => {
    |                                                           ^^^^^^^^^^^^^^^^ variant or associated item not found in `dom::bindings::codegen::InheritTypes::AudioNodeTypeId`
```

I've searched the repo for AudioNodeTypeId, but I can't figure out where I need to add StereoPannerNode in order to get rid of this error?

File: components/script/dom/audionode.rs [hunk 3]
```diff
@@ -280,6 +285,11 @@ impl AudioNodeMethods for AudioNode {
                     return Err(Error::NotSupported);
                 }
             },
+            EventTargetTypeId::AudioNode(AudioNodeTypeId::StereoPannerNode) => {
```
Comment (Manishearth) [inline components/script/dom/audionode.rs]:
```suggestion
            EventTargetTypeId::AudioNode(AudioNodeTypeId::AudioScheduledSourceNode(AudioScheduledSourceNodeTypeId::StereoPannerNode))  => {
```

File: components/script/dom/audionode.rs [hunk 4]
```diff
@@ -236,6 +236,11 @@ impl AudioNodeMethods for AudioNode {
...
```
Comment (Manishearth) [inline components/script/dom/audionode.rs]:
```suggestion
            EventTargetTypeId::AudioNode(AudioNodeTypeId::AudioScheduledSourceNode(AudioScheduledSourceNodeTypeId::StereoPannerNode))  => {
```

**Review by Manishearth — COMMENTED** [review]
this gets subclassed twice, so the type id is a bit more complex

**Comment by PurpleHairEngineer** [comment]
@Manishearth This builds without issues for me now and I ran `rustfmt` on the relevant files. Could you please take a look?

File: components/script/dom/stereopannernode.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        )?;
+        let node_id = source_node.node().node_id();
+        let pan = AudioParam::new(
+            window,
+            context,
+            node_id,
+            ParamType::Pan,
+            AutomationRate::A_rate,
+            *options.pan,
+            f32::MIN,
```
Comment (Manishearth) [inline components/script/dom/stereopannernode.rs]: min is `-1.`, max is `1.`

**Review by Manishearth — COMMENTED** [review]
Looks good!

Now we need to update test expectations. Please run

```
./mach test-wpt /webaudio/the-audio-api --log-raw log
./mach update-wpt log
```

and commit the changes

File: components/script/dom/stereopannernode.rs (hunk 5 above)
Comment (PurpleHairEngineer) [reply components/script/dom/stereopannernode.rs]: @Manishearth Are you saying I need to change this? This is the formatting in gainnode.rs,

Comment (Manishearth) [reply components/script/dom/stereopannernode.rs]: Yes, from the spec the pan parameter is clamped between these values

Comment (PurpleHairEngineer) [reply components/script/dom/stereopannernode.rs]: @Manishearth Okay, I changed it to `-1::MIN` and did the same for MAX. However, when I run `mach.bat test-wpt /webaudio/the-audio-api --log-raw log`, it says `Binary path C:\Users\Maria\servo\target\debug\servo.exe does not exist`.

**Comment by Manishearth** [comment]
There shouldn't be any `MIN`, just `-1.`.

You'll need to wait for a full build before running tests.

**Comment by PurpleHairEngineer** [comment]
@Manishearth When I have `f32::-1.`, `rustfmt` complains: 

```
C:\Users\Maria\servo>rustfmt components/script/dom/stereopannernode.rs
error: expected identifier, found `-`
  --> \\?\C:\Users\Maria\servo\components\script\dom\stereopannernode.rs:63:18
   |
63 |             f32::-1.,
   |                  ^ expected identifier
```

**Comment by Manishearth** [comment]
No, _just_ `-1.`. No f32, no max.

**Comment by PurpleHairEngineer** [comment]
@Manishearth After building the project with no errors, I ran `mach.bat test-wpt /webaudio/the-audio-api --log-raw log`. However, I got a bunch of errors that went back so far that I couldn't even find the first one because it was outside the scope of my command prompt window. These were the last few lines though:

```
  Γöé                         at d:\agent\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
  Γöö   19:     0x7ffcd6564034 - BaseThreadInitThunk

  [182/182] No tests running.
[A[KRan 182 tests finished in 119.0 seconds.
  ΓÇó 1 ran as expected. 0 tests skipped.
  ΓÇó 181 tests crashed unexpectedly

```

When I now run `git status`, it lists the `log` file as being an untracked file. But I don't want to add/commit it if it is messed up now.

**Comment by Manishearth** [comment]
Hm, that seems super broken.

No, you shouldn't be committing the log file :smile: . The file is to be fed to `update-wpt` which will make changes that you _can_ submit.

I'll try this out locally.

File: components/script/dom/stereopannernode.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            return Err(Error::NotSupported);
+        }
+        if node_options.count > 2 || node_options.count == 0 {
+            return Err(Error::NotSupported);
+        }
+        let source_node = AudioScheduledSourceNode::new_inherited(
+            AudioNodeInit::StereoPannerNode(options.into()),
+            context,
+            node_options,
+            0, /* inputs */
```
Comment (Manishearth) [inline components/script/dom/stereopannernode.rs]: this should be 1

**Comment by PurpleHairEngineer** [comment]
@Manishearth I fixed the issue you just mentioned, but I am still getting the same issues when I run `mach.bat test-wpt /webaudio/the-audio-api --log-raw log`.

**Comment by Manishearth** [comment]
Hm, there's a weird failure here that shouldn't be related to your PR

@bors-servo try

let's see what a try run tells us

**Comment by Manishearth** [comment]
Alright, there's something weird with the tests, I think the expectations need a refresh. I just fixed it for you since this requires a bit more work to investigate.

At least one of the crashes you were seeing was https://github.com/servo/media/pull/244

The test doesn't seem to think the implementation is correct, but it's by a very small amount so this might be the offset bug we keep seeing.

@bors-servo try- r+

thanks!

**Comment by PurpleHairEngineer** [comment]
@Manishearth Thank you so much for all your help! It looks like you approved the merge but it is still open. Is there anything else I can do for this?

**Comment by Manishearth** [comment]
No, our CI robot will eventually test and merge this

**Comment by PurpleHairEngineer** [comment]
@Manishearth Oh okay, I was just wondering because it has the label awaiting review.

**Comment by Manishearth** [comment]
@bors-servo r+

ah, did it the wrong way

**Pull request merged**